src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -149,134 +203,282 @@ static void bus_reset(void) {
+      turnaround = 0xDU;
+    else if ( SystemCoreClock >= 15000000U )
+      turnaround = 0xEU;
+    else
+      turnaround = 0xFU;
+
+    // Fullspeed depends on MCU clocks, but we will use 0x06 for 32+ Mhz
+    usb_otg->GUSBCFG |= (turnaround << USB_OTG_GUSBCFG_TRDT_Pos);
   }
 }
 
-static void edpt_schedule_packets(uint8_t const epnum, uint8_t const dir, uint16_t const num_packets, uint16_t total_bytes) {
-    USB_OTG_DeviceTypeDef * const dev = DEVICE_BASE;
-    USB_OTG_OUTEndpointTypeDef * const out_ep = OUT_EP_BASE;
-    USB_OTG_INEndpointTypeDef * const in_ep = IN_EP_BASE;
-
-    // EP0 is limited to one packet each xfer
-    // We use multiple transaction of xfer->max_size length to get a whole transfer done
-    if(epnum == 0) {
-      xfer_ctl_t * const xfer = XFER_CTL_BASE(epnum, dir);
-      total_bytes = tu_min16(ep0_pending[dir], xfer->max_size);
-      ep0_pending[dir] -= total_bytes;
-    }
+static tusb_speed_t get_speed(uint8_t rhport)
+{
+  USB_OTG_DeviceTypeDef * dev = DEVICE_BASE(rhport);
+  uint32_t const enum_spd = (dev->DSTS & USB_OTG_DSTS_ENUMSPD_Msk) >> USB_OTG_DSTS_ENUMSPD_Pos;
+  return (enum_spd == DCD_HIGH_SPEED) ? TUSB_SPEED_HIGH : TUSB_SPEED_FULL;
+}
 
-    // IN and OUT endpoint xfers are interrupt-driven, we just schedule them here.
-    if(dir == TUSB_DIR_IN) {
-      // A full IN transfer (multiple packets, possibly) triggers XFRC.
-      in_ep[epnum].DIEPTSIZ = (num_packets << USB_OTG_DIEPTSIZ_PKTCNT_Pos) |
-                              ((total_bytes << USB_OTG_DIEPTSIZ_XFRSIZ_Pos) & USB_OTG_DIEPTSIZ_XFRSIZ_Msk);
+static void set_speed(uint8_t rhport, tusb_speed_t speed)
+{
+  uint32_t bitvalue;
 
-      in_ep[epnum].DIEPCTL |= USB_OTG_DIEPCTL_EPENA | USB_OTG_DIEPCTL_CNAK;
-      // Enable fifo empty interrupt only if there are something to put in the fifo.
-      if(total_bytes != 0) {
-        dev->DIEPEMPMSK |= (1 << epnum);
-      }
-    } else {
-      // A full OUT transfer (multiple packets, possibly) triggers XFRC.
-      out_ep[epnum].DOEPTSIZ &= ~(USB_OTG_DOEPTSIZ_PKTCNT_Msk | USB_OTG_DOEPTSIZ_XFRSIZ);
-      out_ep[epnum].DOEPTSIZ |= (num_packets << USB_OTG_DOEPTSIZ_PKTCNT_Pos) |
-                                ((total_bytes << USB_OTG_DOEPTSIZ_XFRSIZ_Pos) & USB_OTG_DOEPTSIZ_XFRSIZ_Msk);
+  if ( rhport == 1 )
+  {
+    bitvalue = ((TUSB_SPEED_HIGH == speed) ? DCD_HIGH_SPEED : DCD_FULL_SPEED_USE_HS);
+  }
+  else
+  {
+    bitvalue = DCD_FULL_SPEED;
+  }
 
-      out_ep[epnum].DOEPCTL |= USB_OTG_DOEPCTL_EPENA | USB_OTG_DOEPCTL_CNAK;
-    }
+  USB_OTG_DeviceTypeDef * dev = DEVICE_BASE(rhport);
+
+  // Clear and set speed bits
+  dev->DCFG &= ~(3 << USB_OTG_DCFG_DSPD_Pos);
+  dev->DCFG |= (bitvalue << USB_OTG_DCFG_DSPD_Pos);
 }
 
+#if defined(USB_HS_PHYC) && TUD_OPT_HIGH_SPEED
+static bool USB_HS_PHYCInit(void)
+{
+  USB_HS_PHYC_GlobalTypeDef *usb_hs_phyc = (USB_HS_PHYC_GlobalTypeDef*) USB_HS_PHYC_CONTROLLER_BASE;
+
+  // Enable LDO
+  usb_hs_phyc->USB_HS_PHYC_LDO |= USB_HS_PHYC_LDO_ENABLE;
```

Comment:
Which documentation have you been used for referencing USBPHYC LDO register? I was checking with ST's [RM0431](https://www.st.com/resource/en/reference_manual/dm00305990-stm32f72xxx-and-stm32f73xxx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) and can only see a `LDO_DISABLE` bit.

Reply:
The header have some inconsistant definition. Seee
https://community.st.com/s/question/0D53W00000BMA8HSAX/stm32f7xxh-strange-usbhsphycldoenable-definition

Reply:
Thanks @UweBonnes forwarding this to ST community. If I trust in TRM this bit will disable the LDO. But if it's disabled instead of enabled why don't we get stuck in while loop when waiting for it to be ready? I hope ST can clearify this.

Reply:
I have not many question answered and hints noticed on the ST community forum. So my hope is not high.

Reply:
Yeah, seems they are not going to answer this question. I've checked the F723 discovery board schematic and could not find external power supply for HS peripheral. I don't have this hardware to test the LDO but if HS is working with this code the LDO is defenitely enabled since there is no exernal source provided by this board. We should keep it like this for the moment.